src/Components/Server/test/Circuits/RemoteJSDataStreamTest.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.JSInterop;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    public class RemoteJSDataStreamTest
+    {
+        private static readonly TestRemoteJSRuntime _jsRuntime = new(Options.Create(new CircuitOptions()), Options.Create(new HubOptions()), Mock.Of<ILogger<RemoteJSRuntime>>());
+
+        [Fact]
+        public async void CreateRemoteJSDataStreamAsync_CreatesStream()
+        {
+            // Arrange
+            var jsDataReference = Mock.Of<IJSDataReference>();
+
+            // Act
+            var remoteJSDataStream = await RemoteJSDataStream.CreateRemoteJSDataStreamAsync(_jsRuntime, jsDataReference, totalLength: 100, maxBufferSize: 50, maximumIncomingBytes: 10_000);
+
+            // Assert
+            Assert.NotNull(remoteJSDataStream);
+        }
+
+        [Fact]
+        public async void ReceiveData_DoesNotFindStream()
+        {
+            // Arrange
+            var chunk = new byte[] { 3, 5, 6, 7 };
+            var unrecognizedGuid = 10;
+
+            // Act
+            var success = await RemoteJSDataStream.ReceiveData(_jsRuntime, streamId: unrecognizedGuid, chunk, error: null);
+
+            // Assert
+            Assert.False(success);
+        }
+
+        [Fact]
+        public async void ReceiveData_SuccessReadsBackStream()
+        {
+            // Arrange
+            var jsRuntime = new TestRemoteJSRuntime(Options.Create(new CircuitOptions()), Options.Create(new HubOptions()), Mock.Of<ILogger<RemoteJSRuntime>>());
+            var remoteJSDataStream = await CreateRemoteJSDataStreamAsync(jsRuntime);
+            var streamId = GetStreamId(remoteJSDataStream, jsRuntime);
+            var chunk = new byte[] { 3, 5, 7 };
+
+            // Act
+            var success = await RemoteJSDataStream.ReceiveData(jsRuntime, streamId, chunk, error: null);
+
+            // Assert
+            Assert.True(success);
+        }
+
+        [Fact]
+        public async void ReceiveData_WithError()
+        {
+            // Arrange
+            var jsRuntime = new TestRemoteJSRuntime(Options.Create(new CircuitOptions()), Options.Create(new HubOptions()), Mock.Of<ILogger<RemoteJSRuntime>>());
+            var remoteJSDataStream = await CreateRemoteJSDataStreamAsync(jsRuntime);
+            var streamId = GetStreamId(remoteJSDataStream, jsRuntime);
+
+            // Act
+            var success = await RemoteJSDataStream.ReceiveData(jsRuntime, streamId, chunk: null, error: "some error");
+
+            // Assert
+            Assert.False(success);
```

Comment:
Is it worth asserting that the right exception gets raised on the corresponding `PipeReader`? Same with the other error cases below.

Reply:
Hmm, I'm probably missing something but I don't see where we can actually access the exception passed into `_pipe.Writer.CompleteAsync(exception)`. `_pipe.Reader.OnWriterCompleted` [seems to be how, but that's marked obsolete.](https://docs.microsoft.com/en-us/dotnet/api/system.io.pipelines.pipereader.onwritercompleted?view=dotnet-plat-ext-5.0#System_IO_Pipelines_PipeReader_OnWriterCompleted_System_Action_System_Exception_System_Object__System_Object_)